### DISCUSSION TITLE
Polar stroking: new theory and methods for stroking paths

### DISCUSSION THREAD

**Discussion by JimBobSquarePants** [SixLabors/ImageSharp.Drawing#295] [Ideas]
### Motivation 

Almost all of the time we spend drawing polygons is taken up by offsetting and clipping the polygon to the given pen width. To be considered a first-class alternative to System.Drawing and SkiaSharp we must vastly improve the performance.

I've been doing some research in the area, and it appears that people at Nvidia have been working on an incredibly robust and performant solution to the problem of path stroking. 

At the following link they present videos, documents and a CPU based implementation of an algorithm that can also be implemented on the GPU should we ever choose to follow that path (pun intended).

Abstract + Videos
https://dl.acm.org/doi/abs/10.1145/3386569.3392458

Paper
[kilgard2020.pdf](https://github.com/SixLabors/ImageSharp.Drawing/files/12660437/kilgard2020.pdf)

I've attached a copy of the zip file containing an updated project file as the original has device specific paths for the boost library includes that are required for the demonstration program.

[polar-stroking.zip](https://github.com/SixLabors/ImageSharp.Drawing/files/12659991/polar-stroking.zip)

It's going to take a significant amount of effort to implement this in C# and I personally lack the skills to do it so will be looking for a lot of help. 

Thoughts @SixLabors/core 

Input from the community is most welcome.

**Comment by antonfirsov** [comment]
I agree that there is a lot of space for improvement for stroking, but we should be very careful not to jump into conclusions about the superiority of a particular algorithm without having it understood and compared to alternatives. After taking a brief look at the material linked, I don't see any data about how fast does their tessellation algorithm perform compared to offsetting / clipping.

**Reply by JimBobSquarePants** [reply]
There are no direct numbers I could see but the video talks about and demonstrates the extremely high frame rates possible with the approach. The algorithm is described as non-recursive also. 

I can't imagine Nvidia putting in this effort without it being extremely performant. I do see this as a long-term investment. With an A+ CPU strategy in place designed for GPU portability we open the libraries up to areas we could never compete within normally.

Our approach seems to differ from everyone else, the offsetting + boolean clipping seems very heavy handed. For example, AGG uses a Liang-Barsky line clipping algorithm on segments following stroking and seems to be much performant for a CPU based implementation.

Alternatively, we could have a look at what Blend2d is doing as their benchmarks comfortably beat Skia.
https://blend2d.com/performance.html

**Reply by hypeartist** [reply]
blend2d is using runtime low-level code generation (https://github.com/asmjit/asmjit), so I doubt it's an option here

**Reply by hypeartist** [reply]
Additionally you may take a look at modernized AGG impl - https://github.com/tyoma/agge/tree/master/agge
It utilizes parallel execution and intrinsic stuff.

**Reply by JimBobSquarePants** [reply]
>blend2d is using runtime low-level code generation (https://github.com/asmjit/asmjit), so I doubt it's an option here

We could still use the C++ algorithmic approach as a reference though, no?

**Reply by antonfirsov** [reply]
> demonstrates the extremely high frame rates

The complexity of the curves in the demo isn't high. Their algorithm works with parametric curves not tessellated ones (-> big refactors and possible architectural changes). This can be a deep rabbit hole and the realative benefits compared to alternatives are unclear at the moment.

I would approach this problem by trying to collect a list of the possible algorithms by looking at how various other frameworks solve it, research/understand the algorithms and then either look for code to port or consider a reimplementation (often that can be cheaper and lead to more performant code).

> Liang-Barsky line clipping

AFAIK this is a simple algorithm for clipping lines to a rectangular interest area. I'm missing where does this fit into their pipeline.

I wonder how do AGG's, Blend2D's and Skia's stroking pipelines look like starting from having a parametric curve up until the rasterization @hypeartist any idea on that? Ideally we would need a fast (sort-of) tesselation algorithm that outputs a polygon we can feed to the existing rasterizer, but we can look at other approaches if they are feasible.

**Reply by hypeartist** [reply]
Can't tell much about b2d or skia, but as for agg here some details you may be interested in:
https://www.haiku-os.org/documents/dev/painter_and_how_agg_works

**Reply by hypeartist** [reply]
In agg it's very simple to stroke, say a circle, with a dashed line which every dash segment would be also stroked. You can also insert some transformation at any pipeline stage so, say, apply some distorting transformation to dashed strokes .

**Reply by hypeartist** [reply]
If someone is interested below is results of a simple benchmark (naive C# port of AGG, latest SkiaSharp, and System.Drawing). 
Drawing 20000 stroked circles of random radius, random fill (color + alpha), and random stroke color.

Win11, .NET8, x64

1. AGG (C#) 			- 530 ms
2. SkiaSharp 			- 342 ms
3. System.Drawing 	- 730 ms

![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/3519388/0c9d10f5-3893-4fe7-91a2-6c61f61fe9ac)
![skia](https://github.com/SixLabors/ImageSharp.Drawing/assets/3519388/cfbaec93-4cc1-4196-bc43-9548d28217a2)
![gdiplus](https://github.com/SixLabors/ImageSharp.Drawing/assets/3519388/050b3260-ff24-4f18-9a21-351df741e437)

**Reply by JimBobSquarePants** [reply]
@hypeartist Very interested. That's some really impressive numbers! I'm assuming there's no hardware acceleration on the blending for your AGG code also?

@antonfirsov What's really got me interested in the polar stroking approach as much as anything is the fact that it states it's easily portable to the GPU. I'd love to see a world where ImageSharp.Drawing could be considered something Maui and others could utilize. Toggling between polar stroking and uniform parametric tessellation did seem to show uniform to be faster though when running the demo app on my machine so for the CPU it might not be a great investment 🤔

All in all, I'm really looking for guidance. This is an area that I know far too little about. If the best plan is to borrow bits from @hypeartist port and combine it with our own best processes, then I'm all up for that.

**Reply by antonfirsov** [reply]
> portable to the GPU. I'd love to see a world where ImageSharp.Drawing could be considered something Maui and others could utilize.

I think GPU interop should be a separate discussion, since it leads very far with the questions it opens and may have a deep impact on the API. The thing is that GPU cannot render into system memory buffers directly, it works with GPU memory. If we want the result in an `Image<T>` (eg. in web use cases) we would have to copy the buffers back and forth which (depending on the HW configuration and the complexity of the specific drawing operation) can end up being more expensive than slower CPU rendering. If we want the result to be presented on the screen directly (desktop, mobile use cases), we should avoid these copies meaning that don't need `Image<TPixel>` in the form it exists in the current API; we would rather need a Canvas-like abstraction that allows drawing to various targets.

**Reply by antonfirsov** [reply]
> Our approach seems to differ from everyone else, the offsetting + boolean clipping seems very heavy handed.

I think this is half-true. The stroke-to-fill conversion by offsetting seems to be a pretty common option in other libraries:
https://github.com/google/skia/blob/b961fc35371508346c7552cb8f79651f154f10c3/src/core/SkStroke.cpp#L1419
https://github.com/fontforge/fontforge/pull/3991
https://github.com/nical/lyon/wiki/Stroke-tessellation

The boolean clipping (union) seems to be unnecessary though: naively I would think that drawing the offsetted shapes with the nonzero intersection rule should be enough to deal with (self) intersections. @tocsoft any thoughts here? Am I missing something?

If we could get rid of the clipping, and optimize the offsetter (much of the floating point math there seems to be vectorizable) we could close the path-stroking perf gap with other CPU renderers in a realistic and timely manner. (And again, GPU-enabling ImageSharp.Drawing is a *much* bigger topic.)

**Reply by JimBobSquarePants** [reply]
@antonfirsov I'll have to take your word there as I lack the expertise. 

I'm assuming it would be something like System.Drawings Graphics type or SkiaSharp's [SKCanvas](https://learn.microsoft.com/en-us/dotnet/api/skiasharp.skcanvas?view=skiasharp-2.88)?

That doesn't seem unreasonable, I think in the next major version of the library and should be something to carefully consider to ensure the libraries remain competitive.  

In the interim I wonder how much work it would be to integrate some of the approaches taken in the AGG port in the current API? @hypeartist are you familiar with our inner workings?

**Reply by JimBobSquarePants** [reply]
Sorry, was writing my reply at the same time as your follow up.

**Reply by antonfirsov** [reply]
Regardless of my previous comment, I'm also curious about AGG-s approach. Took a look at the article linked from @hypeartist 's
https://github.com/SixLabors/ImageSharp.Drawing/discussions/295#discussioncomment-7055945 -- it doesn't go into details re stroking much, but looks like AGG is also doing some sort of stroke-to-fill.

**Reply by hypeartist** [reply]
Can you, guys, push me into where inside ISD the actual drawing impl is?

**Reply by JimBobSquarePants** [reply]
Absolutely!

All line stroking falls through the [`OutlinePathExtensions`](https://github.com/SixLabors/ImageSharp.Drawing/blob/dfae51caf7e80d8a8029fa4ef02751d7193a8743/src/ImageSharp.Drawing/Shapes/OutlinePathExtensions.cs) class. The extensions are called via `Pen` instances  If you follow that through you will see [`ClipperOffset`](https://github.com/SixLabors/ImageSharp.Drawing/blob/dfae51caf7e80d8a8029fa4ef02751d7193a8743/src/ImageSharp.Drawing/Shapes/PolygonClipper/ClipperOffset.cs), [`PolygonOffsetter`](https://github.com/SixLabors/ImageSharp.Drawing/blob/dfae51caf7e80d8a8029fa4ef02751d7193a8743/src/ImageSharp.Drawing/Shapes/PolygonClipper/PolygonOffsetter.cs), and [`PolygonClipper`](https://github.com/SixLabors/ImageSharp.Drawing/blob/dfae51caf7e80d8a8029fa4ef02751d7193a8743/src/ImageSharp.Drawing/Shapes/PolygonClipper/PolygonClipper.cs) are used to offset the path and clip it. [`PolygonScanner`](https://github.com/SixLabors/ImageSharp.Drawing/blob/dfae51caf7e80d8a8029fa4ef02751d7193a8743/src/ImageSharp.Drawing/Shapes/Rasterization/PolygonScanner.cs) performs the rasterization.

**Reply by JimBobSquarePants** [reply]
@hypeartist Did you ever have a look at this? If so, did you see anything obvious in our implementation that you would suggest we look at? 

I'm looking at your AGG code just now and not sure how portable some of it is as it's all a bit beyond me.

**Reply by JimBobSquarePants** [reply]
@hypeartist Do you have your sample code that you used to benchmark available. I'd like to start at the top and work down.

**Comment by hypeartist** [comment]
> I'm assuming there's no hardware acceleration on the blending for your AGG code also

No such things in there atm. So I think there is quite some room for optimization possibilities.

**Comment by hypeartist** [comment]
Hey, James (both of you, actually). I'm sorry for being unresponsive. The
end of the year was kinda hell busy at work. And now I'm on NY holidays and
away from keyboard for a week or so. If you can wait I will provide you
with any examples needed next week.

пн, 1 янв. 2024 г. в 06:17, James Jackson-South ***@***.***>:

**Reply by JimBobSquarePants** [reply]
No problem at all!! 😀

**Comment by hypeartist** [comment]
@JimBobSquarePants Hi there!
Here is the demo solution I've used for benching AGG vs SKIA vs GDIP. Clicking on the screen will iterate though engines showing message box with timings. AGG will be rendered directly to the screen while SKIA and GDIP to files. (Since the initially the app is AGG-centric it would take some effort to render to the screen anything other than AGG buffer and I was lazy)
[AggSharp_Test.zip](https://github.com/SixLabors/ImageSharp.Drawing/files/13870861/AggSharp_Test.zip)

**Reply by JimBobSquarePants** [reply]
Thanks for this @hypeartist !! It looks like `ConvStroke` is the key here. If we can use a version of this instead of our current outliner we should be able to avoid most of the work we're doing now with the clipper port and continue to use our existing fast rasterizer. 

It's going to take me a few weeks just to unwind this at best so if you, @antonfirsov, or anyone else reading this fancies having a crack please let me know. 

![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/1da7fef5-df27-42f7-8308-66d339fedde8)

**Reply by hypeartist** [reply]
Don't know if it could help, but still
![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/3519388/05d7a38f-0fa4-4768-8f15-a0ceb60b076f)

**Reply by JimBobSquarePants** [reply]
Yeah thanks, I've got the fundamentals in my head of how it works. It's really more about time constraints and what happens when I mess up the port (which I will). I've got so much to do to ship the next version and I don't know how I'll manage it.

**Reply by antonfirsov** [reply]
I tried the demo and the AggSharp code seems to be the slowest.
The AggSharp test seems to be the slowest, but idk what is being benchmarked. As a next setep IMO someone needs to benchmark only the tessellation code (illustrated above) on it's own, and compare it to PolygonClipper's offsetter, and acquire a good understanding of the algorithm.

Looking at the implementation, it looks inefficient and prone to accumulate errors with the differential increments it's making with tesselation steps. Eg. `Ellipse.cs`:

```C#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public override PathCommand Vertex(ref double x, ref double y)
{
	if (_step == _num)
	{
		++_step;

		return PathCommand.EndPoly | (PathCommand)(PathFlags.Close | PathFlags.Ccw);
	}

	if (_step > _num)
		return PathCommand.Stop;

	var angle = _step / (double)_num * Constants.Misc.PiMul2;
	if (_cw)
		angle = Constants.Misc.PiMul2 - angle;

	x = _x + Math.Cos(angle) * _rx;
	y = _y + Math.Sin(angle) * _ry;
	_step++;

	return _step == 1 ? PathCommand.MoveTo : PathCommand.LineTo;
}
```

**Reply by JimBobSquarePants** [reply]
This is the kind of insight I simply do not have. 😔

**Comment by JimBobSquarePants** [comment]
The more I look into this the more I see the sheer scope of the work involved. It seems more like a complete rewrite which I cannot do.

@antonfirsov Is there a reason why we don't split the work done by `PolygonScanner` into batches?

**Reply by antonfirsov** [reply]
What do you mean by splitting its work into baches? `PolygonScanner` takes a tessellated polygon, and is no bottleneck in the current pipeline. The problem is with the tessellation/extrusion algorithm impemented by `PolygonClipper`.

**Reply by JimBobSquarePants** [reply]
Sorry, I should have been clearer. 

What I mean is to batch operating on the output. Currently it operates like the following: 

```
while (scanner.MoveToNextPixelLine()) 
{
    // Perform operation.
   applicator.Apply(scanline, minX, y);
}
```

Given that our brush applicators are now threadsafe I wonder if we would be able to split that loop into batches. 

The reason I ask this is that for a Fill operation against a complex polygon we are still 3x slower than System.Drawing and 6x slower than SkiaSharp.

|        Method |     Mean |    Error |   StdDev | Ratio | RatioSD |
|-------------- |---------:|---------:|---------:|------:|--------:|
| SystemDrawing | 253.3 us |  2.50 us |  1.95 us |  2.26 |    0.03 |
|    ImageSharp | 766.4 us | 13.54 us | 16.12 us |  6.84 |    0.14 |
|     SkiaSharp | 112.5 us |  1.25 us |  1.17 us |  1.00 |    0.00 |

Benchmark taken from https://github.com/SixLabors/ImageSharp.Drawing/pull/290

Our outline algorithm is still by far the greatest performance bottleneck though once we start drawing as you've identified.

**Reply by antonfirsov** [reply]
I assume you primarily want this to enable parallelization. I think it's not impossible, but quite complicated in this case and would make the code much more complex. Also, we should always remember that gains from algorithm-level parallelizations are cancelled out (or make things worse) in high-load (=already parallel) web scenarios.

I would first profile various operations to identify the current bottlnecks.

**Reply by JimBobSquarePants** [reply]
Yeah, I thought while things like solid brush are fast, slower brushes could do with some help. If it's too complicated then it's not worth the effort then. 

I have been profiling actually though and we get a nice win for free when moving to .NET 8. 

| Method        | Mean     | Error   | StdDev  | Ratio | RatioSD |
|-------------- |---------:|--------:|--------:|------:|--------:|
| SystemDrawing | 254.9 us | 2.87 us | 2.40 us |  2.15 |    0.02 |
| ImageSharp    | 659.4 us | 4.74 us | 3.70 us |  5.57 |    0.04 |
| SkiaSharp     | 118.4 us | 0.75 us | 0.67 us |  1.00 |    0.00 |

Previously `NumericsUtilities.AddToAllElements` was our bottleneck...

<img  alt="image" src="https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/79a2b914-9e86-4f07-b17b-437f606b5793">

... but the new intrisics work in .NET 8 makes that disappear.

<img width="600" alt="image" src="https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/b5e9d5b6-142c-454b-b395-ff9fa802454e">

Our remaining bottlenecks are `ActiveEdgeList.ScanOddEven` and `RasterizerExtensions.ScanCurrentSubpixelLineInto`

**Reply by antonfirsov** [reply]
It might be possible to squeeze a bit more out of `ScanOddEven` by walking the active edges in SIMD register-sized batches and doing FMA for `GetX`. A similar batched loop technique might be applicable for `ScanCurrentSubpixelLineInto` as well.

**Reply by JimBobSquarePants** [reply]
@antonfirsov Do you have any idea what would cause this ghosting? 

I hacked together a copy of the stroke generator and wired it up to the outliner. It does seem to be fast, very fast though...

![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/efa7f353-21ad-401d-ba3b-47ef817eaf74)

My hack is here. https://github.com/SixLabors/ImageSharp.Drawing/tree/4561a3c19e33785422c69545b98806ea3fed4d27

**Reply by tocsoft** [reply]
That sort of ghosting has the same look of an issue in the rasterizer if it fails to account for passing thru vertices/collinear segments, double counting intersections, making it think for that scan line that outside is inside etc, however as you haven't changed the rasterizer that's unlikely. (if memory servers usually causes ghosting all the way to the edge of the image and not just inside the shape , but that might be an even odd vs less than zero difference too).

Another option could be the stroke generator is somehow leaving tiny gaps between segments in some cases, that would probably account for not extending past the bounds of the image too.

**Reply by JimBobSquarePants** [reply]
Thanks @tocsoft 

Sorry I should have tagged you before. How would you approach debugging this?

**Reply by tocsoft** [reply]
I would scale up the outlines (post outlining) i.e. generate outline to a path with thickness 1 then scale up the outlie so the you can then fill it (rather than draw it) to see the differences and any gaps or intersections that are being generated.

```c#
[Theory]
[WithSolidFilledImages(400 * 3, 400 * 3, "Black", PixelTypes.Rgba32, 3)]
[WithSolidFilledImages(400 * 5, 400 * 5, "Black", PixelTypes.Rgba32, 5)]
[WithSolidFilledImages(400 * 10, 400 * 10, "Black", PixelTypes.Rgba32, 10)]
public void LargeGeoJson_Mississippi_LinesScaled(TestImageProvider<Rgba32> provider, int scale)
{
    string jsonContent = File.ReadAllText(TestFile.GetInputFileFullPath(TestImages.GeoJson.States));

    FeatureCollection features = JsonConvert.DeserializeObject<FeatureCollection>(jsonContent);

    Feature missisipiGeom = features.Features.Single(f => (string)f.Properties["NAME"] == "Mississippi");

    Matrix3x2 transform = Matrix3x2.CreateTranslation(-87, -54)
                    * Matrix3x2.CreateScale(60, 60);
    IReadOnlyList<PointF[]> points = PolygonFactory.GetGeoJsonPoints(missisipiGeom, transform);

    using Image<Rgba32> image = provider.GetImage();
    var pen = new SolidPen(new SolidBrush(Color.White), 1.0f);
    foreach (PointF[] loop in points)
    {
        IPath outline = pen.GeneratePath(new Path(loop).Transform(Matrix3x2.CreateTranslation(0.5F, 0.5F)));
        outline = outline.Transform(Matrix3x2.CreateScale(scale, scale));
        image.Mutate(c => c.Fill(pen.StrokeFill, outline));
    }

    // Strict comparer, because the image is sparse:
    ImageComparer comparer = ImageComparer.TolerantPercentage(0.0001F);

    string details = $"Scale({scale})";
    image.DebugSave(provider, details, appendPixelTypeToFileName: false, appendSourceFileOrDescription: false);
    image.CompareToReferenceOutput(comparer, provider, testOutputDetails: details, appendPixelTypeToFileName: false, appendSourceFileOrDescription: false);
}
```
### master vs your branch.
![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/166440/63cbeb4a-106e-446b-ab76-638c6715cbed)

There are a lot of self intersections going on here which could be related, the 'dirtiness' of the generated outline (probably the main reason clipper is slower, as it fixes this stuff up) might actually just be uncovering some pre-existing issues/edge cases in our rasterizer, if I had to guess it might be holes being particularly small and thus triggering rounding issues with floats.

Additionally it looks like there is difference in that last line segment too, looks like clipper is closing the path where as the new code isn't.

**Reply by JimBobSquarePants** [reply]
Thanks @tocsoft 

I believe that the lack of cleanup of self-intersection within the AGG stroker is highlighting a pre-existing issue in `PolygonScanner` https://github.com/SixLabors/ImageSharp.Drawing/issues/106 I'd appreciate any assistance that could be offered solving that one. It is interesting to note

Regarding any lack of closing. My naive hack was chopping off the last input point, but I cannot get this damn thing to correctly close the path. It's driving me to despair!

![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/e4c0a1b0-1cb1-4331-b0f8-4aab1422d076)

**Reply by JimBobSquarePants** [reply]
OK.....

So, I figured out how to close the polygon. AGG leaves that to the rasterizer, passing a path command telling it to draw a line to the intitial point. 

![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/675e312a-4ed6-49ba-9bb3-b967d6589716)

Benchmarks are incredibly promising. I imagine we can squeeze more off this but we're looking at a massive speedup.
If I can figure out what causes the ghosting in the robustness tests, then we could be on to a winner. 

#### DrawPolygonMediumThick
 
Main

|        Method |      Mean |     Error |    StdDev | Ratio | RatioSD |
|-------------- |----------:|----------:|----------:|------:|--------:|
| SystemDrawing |  1.352 ms | 0.0094 ms | 0.0083 ms |  0.92 |    0.01 |
|    ImageSharp | 14.104 ms | 0.2779 ms | 0.3089 ms |  9.61 |    0.25 |
|     SkiaSharp |  1.473 ms | 0.0068 ms | 0.0053 ms |  1.00 |    0.00 |

AGG Hack

| Method        | Mean     | Error     | StdDev    | Ratio | RatioSD |
|-------------- |---------:|----------:|----------:|------:|--------:|
| SystemDrawing | 1.369 ms | 0.0102 ms | 0.0091 ms |  0.93 |    0.01 |
| ImageSharp    | 3.589 ms | 0.0380 ms | 0.0337 ms |  2.44 |    0.03 |
| SkiaSharp     | 1.475 ms | 0.0114 ms | 0.0089 ms |  1.00 |    0.00 |

**Reply by hypeartist** [reply]
@JimBobSquarePants Don't know if it matters much, but those renderings above have different line joins: left -  bevel join, right - miter join.

**Reply by JimBobSquarePants** [reply]
Yeah, I just haven't wired up the joins yet. Using the default while I try to get the basics working.

**Reply by JimBobSquarePants** [reply]
@antonfirsov @tocsoft Adjusting the `GraphicsOptions.AntialiasSubpixelDepth` to a minimum of 34 makes the ghosting issue go away. I don't know whether this is a feature or a bug.

**Reply by antonfirsov** [reply]
> GraphicsOptions.AntialiasSubpixelDepth to a minimum of 34 makes the ghosting issue go away

My guess is that taking more samples over a polygon with tessellation issues (especially around self intersections) makes the ghost lines (possible impact from #106) weigh less in the rasterization and become innoticable.

Note that #106 could be improved by switching the rasterizer precision to `double`, but that would penalize rendering performance for all workloads not just line drawing. Also, self-intersections are rendered incorrectly even without ghost lines as illustrated in https://github.com/SixLabors/ImageSharp.Drawing/discussions/295#discussioncomment-8265438.

We need to figure out how to get this right. The questions I would ask at this point: is there a quality gap between the original C++ AGG polygon stroker and the C# port? How does AGG rendering handle self-intersections end-to end?

**Reply by JimBobSquarePants** [reply]
@antonfirsov The C# stroker uses double precision and appears to be a straight port. It actually does a better job than clipper for examples like our triangles above as it does not introduce self-intersections around the joints. For dashed lines it still requires clipping which I've wired up in my branch though I'm curious to investigate how AGG does dashes as it would be nice to avoid the clipping. 

There are two clippers in AGG. There's a general polygon clipper and the Liang-Barsky line clipper. The rasterizer uses the latter for clipping lines to the area of interest, the former is strictly opt-in. Looking at type `RasterizerCellsAA` in the port @hypeartist has implemented it appears that the rasterizer is using 256 subsamples. I'm struggling to follow exactly how it remains performant though if that is the case.

**Reply by hypeartist** [reply]
@JimBobSquarePants @antonfirsov  Don't know if you, guys, have seen this or not, but still:
https://gasiulis.name/parallel-rasterization-on-cpu/ (https://github.com/aurimasg/blaze/tree/main)

**Reply by hypeartist** [reply]
> is there a quality gap between the original C++ AGG polygon stroker and the C# port

It's almost 1-1 port, so I assume there shouldn't be any 'quality gap'

**Reply by JimBobSquarePants** [reply]
Oooh! That does look interesting. I’m no C++ guy though so would struggle with a port if it could be integrated.

**Reply by hypeartist** [reply]
<details>
<summary>@JimBobSquarePants Just a bit of show-off:</summary>
<img src="https://github.com/SixLabors/ImageSharp.Drawing/assets/3519388/00ce5d90-dafa-4b99-90c2-db477192bd10"/>
</details>

**Reply by JimBobSquarePants** [reply]
I'm going to need a lot of help. 2.5x slower than System.Drawing on my machine plus I still haven't figured out how to fix the ghosting. I really need to try porting that other rasterizer. 😔

```
BenchmarkDotNet v0.13.10, Windows 11 (10.0.22631.3085/23H2/2023Update/SunValley3)
11th Gen Intel Core i7-11370H 3.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.200-preview.23624.5
  [Host]   : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
  ShortRun : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3

| Method         | Mean    | Error    | StdDev   |
|--------------- |--------:|---------:|---------:|
| DrawImageSharp | 1.855 s | 0.0426 s | 0.0023 s |
```

![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/aa59c773-ea84-4be5-a968-bcbd15d7ac3c)
